pkg/cac/cac.go

```diff
@@ -8,85 +8,88 @@ import (
 	"bytes"
 	"encoding/binary"
 	"errors"
+	"fmt"
 
 	"github.com/ethersphere/bee/pkg/bmtpool"
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
 var (
-	errTooShortChunkData = errors.New("short chunk data")
-	errTooLargeChunkData = errors.New("data too large")
+	ErrChunkDataShort = errors.New("chunk data length is not sufficient")
+	ErrChunkDataLarge = errors.New("chunk data exceeds maximum allowed length")
 )
 
 // New creates a new content address chunk by initializing a span and appending the data to it.
 func New(data []byte) (swarm.Chunk, error) {
 	dataLength := len(data)
-	if dataLength > swarm.ChunkSize {
-		return nil, errTooLargeChunkData
-	}
 
-	if dataLength == 0 {
-		return nil, errTooShortChunkData
+	if err := validateDataLength(dataLength); err != nil {
+		return nil, fmt.Errorf("invalid CAC data length (%d): %w", dataLength, err)
 	}
 
 	span := make([]byte, swarm.SpanSize)
 	binary.LittleEndian.PutUint64(span, uint64(dataLength))
+
 	return newWithSpan(data, span)
 }
 
 // NewWithDataSpan creates a new chunk assuming that the span precedes the actual data.
 func NewWithDataSpan(data []byte) (swarm.Chunk, error) {
 	dataLength := len(data)
-	if dataLength > swarm.ChunkSize+swarm.SpanSize {
-		return nil, errTooLargeChunkData
-	}
 
-	if dataLength < swarm.SpanSize {
-		return nil, errTooShortChunkData
+	if err := validateDataLength(dataLength - swarm.SpanSize); err != nil {
+		return nil, fmt.Errorf("invalid CAC data length (%d): %w", dataLength, err)
 	}
+
 	return newWithSpan(data[swarm.SpanSize:], data[:swarm.SpanSize])
 }
 
+// validateDataLength validates if data length (without span) is correct.
+func validateDataLength(dataLength int) error {
+	if dataLength <= 0 { // dataLength could be negative when span size is subtracted
```

Comment:
In particular, consider the following chunk from mainnet which is used by my OSM dataset.
b34ca8c22b9e982354f9c7f50b470d66db428d880c8a904d5fe4ec9713171526

Reply:
The comment doesn't match the condition check.  It's not invalidating negatives, but also invalidating zero.

Reply:
The reason why its invalidating zero data lengths was explained above.

Reply:
> In particular, consider the following chunk from mainnet which is used by my OSM dataset. b34ca8c22b9e982354f9c7f50b470d66db428d880c8a904d5fe4ec9713171526

Just to re-verifiy, I uploaded a zero length file to the gateway that resulting in the following reference.   If you expand the manifest recursively on that reference, you will find that it resolves to the previous zero-length reference for the file's (zero length) content.

4974f1792f764351cb64b72412f536f6721855f9535e7c31c7b308b20e32f43d